### PR TITLE
Add prepare_dlc_dev_environment script to assist in preparation of dev toml

### PR DESCRIPTION
**Pull request by jessepeq23**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
- Add src/prepare_dlc_dev_environment.py, which allows a user to automatically format toml file for pytorch/tensorflow builds and tests
- Added new quick checks unit tests to test the functionality of the script

### Tests run
- Execute unit tests

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**
<details>
<summary>Expand</summary>

- [ ] `sagemaker_remote_tests = true`
- [ ] `sagemaker_efa_tests = true`
- [ ] `sagemaker_rc_tests = true`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

</details>

### Formatting
- [x] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

#### Builds to Execute
<details>
<summary>Expand</summary>

Fill out the template and click the checkbox of the builds you'd like to execute

*Note: Replace with <X.Y> with the major.minor framework version (i.e. 2.2) you would like to start.*

- [ ] build_pytorch_training_<X.Y>_sm
- [ ] build_pytorch_training_<X.Y>_ec2

- [ ] build_pytorch_inference_<X.Y>_sm
- [ ] build_pytorch_inference_<X.Y>_ec2
- [ ] build_pytorch_inference_<X.Y>_graviton

- [ ] build_tensorflow_training_<X.Y>_sm
- [ ] build_tensorflow_training_<X.Y>_ec2

- [ ] build_tensorflow_inference_<X.Y>_sm
- [ ] build_tensorflow_inference_<X.Y>_ec2
- [ ] build_tensorflow_inference_<X.Y>_graviton
</details>

### PR Checklist 
<details>
<summary>Expand</summary>

- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

#### NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ec2_benchmark_tests = true` or `sagemaker_benchmark_tests = true`
</details>

### Pytest Marker Checklist
<details>
<summary>Expand</summary>

- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type
</details>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/prepare_dlc_dev_environment.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        Write a function that, given a list of frameworks, assembles a dictionary with key/value pairs:
+        {"build_frameworks": ["framework1", "framework2"]}
+
+        Make sure there are no repeats
+
+        frameworks is a list
+        return a dictionary object
+        """
+        unique_frameworks = list(set(frameworks))
+        self._overrides["build_frameworks"] = unique_frameworks
```
Comment (arjkesh) [inline src/prepare_dlc_dev_environment.py:57]: we should actually be setting this to self._overrides["build"]["build_frameworks"]

File: src/prepare_dlc_dev_environment.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        """
+        Job type can be one of (or both) "training" or "inference"
+
+        If job_type is training, set build_training to True, and build_inference to False
+        If job type is inference, set build_training to False, and build_inference to True
+        If both are set, set both to true
+        """
+        build_training = "training" in job_types
+        build_inference = "inference" in job_types
+        self._overrides["build_training"] = build_training
```
Comment (arjkesh) [inline src/prepare_dlc_dev_environment.py:69]: we should actually be setting this to self._overrides["build"]["build_training"] and self._overrides["build"]["build_inference"]

File: test/dlc_tests/sanity/quick_checks/test_prepare_dev_env.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+
+@pytest.mark.quick_checks
+@pytest.mark.model("N/A")
+@pytest.mark.integration("build_frameworks")
+def test_build_frameworks():
+    overrider = prepare_dlc_dev_environment.TomlOverrider()
+    overrider.set_build_frameworks(("pytorch", "tensorflow"))
+    
+    assert overrider.overrides == {"build_frameworks": ["pytorch", "tensorflow"]}
```
Comment (arjkesh) [inline test/dlc_tests/sanity/quick_checks/test_prepare_dev_env.py:13]: you'll need to update your tests as well with the new expected paths

File: src/prepare_dlc_dev_environment.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    return parser.parse_args()
+
+
+class TomlOverrider:
+    def __init__(self):
+        self._overrides = {"build": {}}
+
+    def set_build_frameworks(self, frameworks):
+        """
+        Write a function that, given a list of frameworks, assembles a dictionary with key/value pairs:
```
Comment (arjkesh) [inline src/prepare_dlc_dev_environment.py:49]: Can you update this docstring so that it includes the description of the method now, rather than the instructions?

File: test/dlc_tests/sanity/quick_checks/test_dlc_developer_config.py [hunk 5]
```diff
@@ -24,7 +24,7 @@ def test_developer_configuration():
     assert config.is_deep_canary_mode_enabled() is False
 
     # Check build settings
-    assert config.parse_dlc_developer_configs("build", "build_frameworks") == []
+    assert config.parse_dlc_developer_configs("build", "build_frameworks") == ["pytorch"]
```
Comment (arjkesh) [inline test/dlc_tests/sanity/quick_checks/test_dlc_developer_config.py:27]: Revert this change, default should still stay as is

File: dlc_developer_config.toml [hunk 6]
```diff
@@ -1,159 +1,74 @@
 [dev]
```
Comment (arjkesh) [inline dlc_developer_config.toml:1]: You can revert this change back to its original state now

**Review by arjkesh — DISMISSED** [review]

**Comment by jessepeq23** [comment]
will be correctly updating the "--tests" on Tuesday (06/10/24)

File: src/prepare_dlc_dev_environment.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+            "eks",
+            "sagemaker_remote",
+            "sagemaker_local",
+        ],
+        default=["ec2", "ecs", "eks", "sagemaker_remote", "sagemaker_local"],
+        help="Types of tests to run",
+    )
+    parser.add_argument(
+        "--dev_mode",
+        choices=["graviton_mode", "neuron_mode", "deep_canary_mode"],
```
Comment (arjkesh) [inline src/prepare_dlc_dev_environment.py:55]: you can prob just do neuronx mode for now

File: src/prepare_dlc_dev_environment.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+                "nightly_pr_test_mode": False,
+                "use_scheduler": False,
+            }
+
+    def set_dev_mode(self, dev_mode):
+        """
+        Set the dev mode based on the user input.
+        Valid choices are 'graviton_mode', 'neuron_mode', and 'deep_canary_mode'.
+        """
+        self._overrides["dev"]["graviton_mode"] = False
```
Comment (arjkesh) [inline src/prepare_dlc_dev_environment.py:142]: remove 3 lines

File: src/prepare_dlc_dev_environment.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        nargs="+",
+        choices=["training", "inference"],
+        default=["training", "inference"],
+        help="Training and inference containers to prepare developer environment",
+    )
+    parser.add_argument(
+        "--tests",
+        nargs="+",
+        choices=[
+            "benchmark",
```
Comment (arjkesh) [inline src/prepare_dlc_dev_environment.py:43]: leave benchmark alone for now

Comment (arjkesh) [reply src/prepare_dlc_dev_environment.py:43]: add sanity

File: test/dlc_tests/sanity/quick_checks/test_prepare_dev_env.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+
+@pytest.mark.quick_checks
+@pytest.mark.model("N/A")
+@pytest.mark.integration("build_frameworks")
+def test_build_frameworks():
+    overrider = prepare_dlc_dev_environment.TomlOverrider()
+    overrider.set_build_frameworks(("pytorch", "tensorflow"))
+
+    assert overrider.overrides == {"build": {"build_frameworks": ["pytorch", "tensorflow"]}}
```
Comment (arjkesh) [inline test/dlc_tests/sanity/quick_checks/test_prepare_dev_env.py:13]: looks like these are going to have to be more specific. you can do something like 

```
assert overrider.overrides["build"]["build_frameworks"] == ["pytorch", "tensorflow"]
```

and apply this same pattern to the other failing tests below, making sure to swap out the keys and values for their expected results

example for job types

```
assert overrider.overrides["build"]["build_training"] == True and overrider.overrides["build"]["build_inference"] == True 
```

**Review by arjkesh — APPROVED** [review]
